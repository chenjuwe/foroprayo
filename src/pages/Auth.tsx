import React, { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import { AuthForm } from '@/components/auth/AuthForm';
import { AuthFooter } from '@/components/auth/AuthFooter';
import { useToast } from '@/hooks/use-toast';
import { useFirebaseAuth } from '@/contexts/FirebaseAuthContext';
import { useFirebaseAuthStore } from '@/stores/firebaseAuthStore';
import { log } from '@/lib/logger';
import { useFirebaseAvatar } from '@/hooks/useFirebaseAvatar';

// 頁面預加載函數 - 這會在模組層級執行，而不是在組件內部
(() => {
  // 預加載關鍵資源
  const preloadResources = [
    '/prayers', // 頁面路由
    '/apple-touch-icon.png', // 關鍵圖片
    '/pwa-192x192.png'
  ];
  
  preloadResources.forEach(resource => {
    if (resource.endsWith('.png')) {
      const img = new Image();
      img.src = resource;
    } else {
      const link = document.createElement('link');
      link.rel = 'prefetch';
      link.href = resource;
      document.head.appendChild(link);
    }
  });
})();

export default function Auth() {
  const navigate = useNavigate();
  const { toast } = useToast();
  const { signIn, signUp, refreshUserAvatar } = useFirebaseAuth();
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [isLogin, setIsLogin] = useState(true);
  const [isLoading, setIsLoading] = useState(false);
  const [isRedirecting, setIsRedirecting] = useState(false);
  
  // 初始化 Firebase 認證狀態
  const initFirebaseAuth = useFirebaseAuthStore(state => state.initAuth);
  const { refreshAvatar } = useFirebaseAvatar();
  
  // 在組件掛載時初始化 Firebase 認證狀態並預加載資源
  useEffect(() => {
    // 初始化認證
    initFirebaseAuth();
  }, [initFirebaseAuth]);
  
  // 超高速跳轉函數 - 使用最快的方式跳轉
  const ultraFastRedirect = (userId?: string) => {
    if (isRedirecting) return; // 防止重複跳轉
    setIsRedirecting(true);
    
    // 使用內聯腳本技術，將會立即執行，比 setTimeout 更快
    const script = document.createElement('script');
    script.textContent = `
      try {
        // 記錄跳轉開始時間 (用於性能監控)
        window.__redirectStart = Date.now();
        
        // 使用 location.replace 提供最快的跳轉（不會增加瀏覽歷史）
        window.location.replace('/prayers');
      } catch(e) {
        // 降級方案
        window.location.href = '/prayers'; 
      }
    `;
    document.body.appendChild(script);
  };

  const handleSubmit = async (e?: React.FormEvent) => {
    e?.preventDefault();
    
    // 防止重複提交
    if (isLoading || isRedirecting) return;
    
    setIsLoading(true);

    try {
      // 首先檢查網絡狀態
      if (!navigator.onLine) {
        throw new Error('網絡連接不可用，請檢查您的連接並重試');
      }

      if (isLogin) {
        // 顯示登入中提示（立即顯示以提供視覺反饋）
        toast({ title: "登入中...", description: "請稍候" });
        
        // 檢查是否存在匹配的緩存用戶
        try {
          const cachedUserString = localStorage.getItem('auth_user');
          if (email && cachedUserString) {
            const cachedUser = JSON.parse(cachedUserString);
            // 檢查緩存的電子郵件是否匹配
            if (cachedUser.email === email && 
                Date.now() - cachedUser.timestamp < 24 * 60 * 60 * 1000) {
              
              // 使用緩存快速登入
              toast({ 
                title: "登入成功", 
                description: "使用緩存認證快速登入" 
              });
              
              // 存儲當前郵箱作為最後使用的用戶
              localStorage.setItem('last_user', email);
              
              // 立即跳轉
              ultraFastRedirect(cachedUser.uid);
              
              // 在後台嘗試實際登入
              setTimeout(async () => {
                try {
                  await signIn(email, password);
                } catch (e) {
                  // 忽略錯誤，因為用戶已經使用緩存登入
                }
              }, 100);
              
              setIsLoading(false);
              return;
            }
          }
        } catch (e) {
          // 忽略緩存處理錯誤
        }
        
        // 使用 Firebase 登入
        try {
          const { user, error, fromCache } = await signIn(email, password);
          
          if (error) {
            // 特別處理資源限制問題
            // if (resourceLimit) {
            //   toast({
            //     title: "使用離線模式登入",
            //     description: "Firebase服務繁忙，使用本地登入"
            //   });
              
            //   // 如果有緩存的用戶且登入成功
            //   if (user) {
            //     // 存儲當前郵箱作為最後使用的用戶
            //     localStorage.setItem('last_user', email);
                
            //     // 立即跳轉
            //     ultraFastRedirect(user.uid);
            //     setIsLoading(false);
            //     return;
            //   }
            // }
            
            throw new Error(error);
          }
          
          if (user) {
            // 處理離線或緩存登入
            if (fromCache) {
              toast({ 
                title: "登入成功", 
                description: "歡迎回來！" 
              });
              
              // 存儲當前郵箱作為最後使用的用戶
              localStorage.setItem('last_user', email);
              
              // 立即跳轉
              ultraFastRedirect(user.uid);
              setIsLoading(false);
              return;
            }
            
            // 正常登入流程
            initFirebaseAuth();
            toast({ title: "登入成功", description: "歡迎回來！" });
            ultraFastRedirect(user.uid);
            
            // 將非關鍵操作延後
            setTimeout(() => {
              try {
                refreshUserAvatar();
              } catch (e) {
                // 忽略錯誤
              }
            }, 100);
          }
        } catch (authError) {
          // 處理特定登入錯誤
          const errorMessage = authError instanceof Error ? authError.message : '登入失敗';
          
          // 再次嘗試緩存登入（作為最後的降級手段）
          try {
            const lastUser = localStorage.getItem('last_user');
            const cachedUserString = localStorage.getItem('auth_user');
            
            // 如果有緩存且電子郵件匹配
            if (cachedUserString && lastUser === email) {
              const cachedUser = JSON.parse(cachedUserString);
              
              toast({
                title: "嘗試使用緩存登入",
                description: "正常登入失敗，使用備用方式"
              });
              
              ultraFastRedirect(cachedUser.uid);
              setIsLoading(false);
              return;
            }
          } catch (e) {
            // 忽略備用方案錯誤
          }
          
          // 徹底失敗，顯示錯誤
          toast({
            title: "登入失敗",
            description: errorMessage,
            variant: "destructive",
          });
        }
      } else {
        // 使用 Firebase 註冊
        log.debug('嘗試 Firebase 註冊', { 
          email: email.substring(0, 3) + '***', // 只記錄部分郵件，保護隱私
          timestamp: Date.now()
        }, 'Auth');
        
        const { user, error } = await signUp(email, password);
        if (error) throw new Error(error);
        
        if (user) {
          log.debug('Firebase 註冊成功', { 
            userId: user.uid, 
            email: user.email
          }, 'Auth');
          
          // 確保 Firebase Auth 狀態更新
          initFirebaseAuth();
          
          // 強制刷新頭像
          refreshUserAvatar();
          
          // 觸發頭像更新事件
          window.dispatchEvent(new CustomEvent('avatar-updated', { 
            detail: { 
              userId: user.uid,
              timestamp: Date.now(),
              source: 'Auth'
            }
          }));
          
          toast({ title: "註冊成功", description: "帳號創建完成，正在為你登入..." });
          
          // 延遲跳轉，給認證狀態更新更多時間
          log.debug('準備導航到 /prayers (註冊後)', { 
            userId: user.uid,
            currentPath: window.location.pathname
          }, 'Auth');
          
          // 簡化導航邏輯，使用 setTimeout + window.location.replace
          setTimeout(() => {
            log.debug('執行導航到 /prayers (註冊後)', { 
              userId: user.uid,
              authState: !!user,
              timestamp: Date.now()
            }, 'Auth');
            
            // 使用 replace 而不是 href，避免瀏覽器歷史記錄問題
            window.location.replace('/prayers');
          }, 1500); // 增加到 1.5 秒
        } else {
          toast({ title: "註冊完成", description: "請檢查你的電子信箱以完成帳號啟用" });
        }
      }
    } catch (error: unknown) {
      const errorMessage = error instanceof Error ? error.message : "請稍後再試";
      
      // 處理配額超出的特殊情況
      if (errorMessage.includes('quota') || 
          errorMessage.includes('AUTH_QUOTA_LIMIT') ||
          errorMessage.includes('認證服務暫時不可用') ||
          errorMessage.includes('INSUFFICIENT_RESOURCES') ||
          errorMessage.includes('Firebase 服務暫時不可用')) {
          
        // 檢查是否有緩存的用戶資訊
        const cachedUser = localStorage.getItem('auth_user');
        if (cachedUser && email) {
          try {
            const userData = JSON.parse(cachedUser);
            
            // 記錄上次使用的電子郵件，便於下次快速緩存恢復
            localStorage.setItem('last_user', email);
            
            toast({
              title: "使用備用登入方式",
              description: "Firebase服務繁忙，使用本地認證"
            });
            
            // 立即跳轉
            ultraFastRedirect(userData.uid);
            setIsLoading(false);
            return;
          } catch (e) {
            // 忽略錯誤並繼續下面的錯誤處理
          }
        }
        
        toast({
          title: "伺服器忙碌中",
          description: "請稍後再試，或使用訪客登入",
          variant: "destructive",
        });
      } else {
        // 一般錯誤處理
        toast({
          title: isLogin ? "登入失敗" : "註冊失敗",
          description: errorMessage,
          variant: "destructive",
        });
      }
    } finally {
      setIsLoading(false);
    }
  };

  const handleGuestAccess = () => {
    try {
      log.debug('嘗試進入訪客模式', { currentPath: window.location.pathname }, 'Auth');
      
      // 設置訪客模式
      localStorage.setItem('guestMode', 'true');
      
      // 顯示訪客登入成功提示
      toast({ title: "訪客登入中...", description: "請稍候" });
      
      // 確保頁面重新導向
      console.log('準備導航到訪客模式 /prayers 頁面');
      
      // 使用 setTimeout 確保 localStorage 設置有時間完成
      setTimeout(() => {
        // 使用直接的導航方式
        window.location.href = '/prayers';
      }, 100);
    } catch (error) {
      console.error('訪客模式導航失敗', error);
      // 降級方案 - 如果發生錯誤，使用最簡單的跳轉
      window.location.href = '/prayers';
    }
  };

  return (
    <div className="min-h-screen flex flex-col" style={{ backgroundColor: '#FFE5D9' }}>
      <div className="flex-1 flex flex-col items-center px-6 py-8">
        <div className="relative z-10 w-full max-w-sm">
          <AuthForm 
            email={email}
            password={password}
            isLogin={isLogin}
            isLoading={isLoading}
            onEmailChange={(e) => setEmail(e.target.value)}
            onPasswordChange={(e) => setPassword(e.target.value)}
            onToggle={() => setIsLogin(!isLogin)}
            onSubmit={handleSubmit}
          />
        </div>
        
        <div 
          style={{ 
            position: 'absolute', 
            top: '544px', 
            left: '50%', 
            transform: 'translateX(-50%)',
            width: '270px',
            zIndex: 20
          }}
        >
          <button
            type="button"
            onClick={() => handleSubmit()}
            disabled={isLoading}
            className="text-black w-full"
            style={{ 
              backgroundColor: '#95d2f4',
              height: '30px',
              borderRadius: '15px',
              fontSize: '14px',
              opacity: isLoading ? 0.7 : 1,
              marginBottom: '40px'
            }}
          >
            {isLoading ? '處理中...' : (isLogin ? '登入帳號' : '註冊帳號')}
          </button>
          <button
            type="button"
            onClick={handleGuestAccess}
            className="text-black w-full"
            style={{ 
              backgroundColor: '#4D9BD9',
              height: '30px',
              borderRadius: '15px',
              fontSize: '14px',
              display: 'flex',
              alignItems: 'center',
              justifyContent: 'center',
            }}
          >
            訪客帳號
          </button>
        </div>
      </div>
    </div>
  );
} 