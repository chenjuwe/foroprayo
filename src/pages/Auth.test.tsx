import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { vi, describe, it, expect, beforeEach } from 'vitest';
import Auth from './Auth';

// Mock React Router
vi.mock('react-router-dom', async () => {
  const actual = await vi.importActual('react-router-dom');
  return {
    ...actual,
    useNavigate: () => vi.fn(),
    useLocation: () => ({ pathname: '/auth' }),
  };
});

// Mock toast
vi.mock('sonner', () => ({
  toast: {
    success: vi.fn(),
    error: vi.fn(),
    warning: vi.fn(),
    info: vi.fn(),
  },
}));

// Mock AuthForm component
vi.mock('@/components/auth/AuthForm', () => ({
  AuthForm: ({ email, password, isLogin, isLoading, onEmailChange, onPasswordChange, onToggle, onSubmit }: any) => (
    <div data-testid="auth-form">
      <div>Auth Form Component</div>
      <input
        type="email"
        value={email}
        onChange={onEmailChange}
        placeholder="電子郵件"
        data-testid="email-input"
      />
      <input
        type="password"
        value={password}
        onChange={onPasswordChange}
        placeholder="密碼"
        data-testid="password-input"
      />
      <button
        type="submit"
        onClick={onSubmit}
        disabled={isLoading}
        data-testid="submit-button"
      >
        {isLogin ? '登入' : '註冊'}
      </button>
      <button
        type="button"
        onClick={onToggle}
        data-testid="toggle-button"
      >
        {isLogin ? '切換到註冊' : '切換到登入'}
      </button>
    </div>
  ),
}));

// Mock Firebase Auth Hook
vi.mock('@/hooks/useFirebaseAuth', () => ({
  useFirebaseAuth: () => ({
    signIn: vi.fn(),
    signUp: vi.fn(),
    refreshUserAvatar: vi.fn(),
  }),
}));

// Mock Firebase Auth Store
vi.mock('@/stores/firebaseAuthStore', () => ({
  useFirebaseAuthStore: (selector: any) => {
    const state = {
      user: null,
      isAuthLoading: false,
      isLoggedIn: false,
      initAuth: vi.fn(),
    };
    return selector ? selector(state) : state;
  },
}));

// Mock Firebase Avatar Hook
vi.mock('@/hooks/useFirebaseAvatar', () => ({
  useFirebaseAvatar: () => ({
    refreshAvatar: vi.fn(),
  }),
}));

// Mock logger
vi.mock('@/lib/logger', () => ({
  log: {
    debug: vi.fn(),
    error: vi.fn(),
    warn: vi.fn(),
    info: vi.fn(),
  },
}));

// Mock localStorage
const localStorageMock = {
  getItem: vi.fn(),
  setItem: vi.fn(),
  removeItem: vi.fn(),
  clear: vi.fn(),
};
Object.defineProperty(window, 'localStorage', {
  value: localStorageMock,
});

// Mock window.location
Object.defineProperty(window, 'location', {
  value: {
    href: '',
  },
  writable: true,
});

const renderAuth = () => {
  return render(
    <BrowserRouter>
      <Auth />
    </BrowserRouter>
  );
};

describe('Auth Page', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('基本渲染', () => {
    it('應該正確渲染認證頁面', async () => {
      await act(async () => {
        renderAuth();
      });
      
      expect(screen.getByTestId('auth-form')).toBeInTheDocument();
    });

    it('應該預設顯示登入模式', async () => {
      await act(async () => {
        renderAuth();
      });
      
      const submitButton = screen.getByTestId('submit-button');
      expect(submitButton).toHaveTextContent('登入');
    });

    it('應該包含所有必要的表單元素', async () => {
      await act(async () => {
        renderAuth();
      });
      
      expect(screen.getByTestId('email-input')).toBeInTheDocument();
      expect(screen.getByTestId('password-input')).toBeInTheDocument();
      expect(screen.getByTestId('submit-button')).toBeInTheDocument();
      expect(screen.getByTestId('toggle-button')).toBeInTheDocument();
    });
  });

  describe('表單交互', () => {
    it('應該正確處理電子郵件輸入', async () => {
      await act(async () => {
        renderAuth();
      });
      
      const emailInput = screen.getByTestId('email-input');
      await act(async () => {
        fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
      });
      
      expect(emailInput).toHaveValue('test@example.com');
    });

    it('應該正確處理密碼輸入', async () => {
      await act(async () => {
        renderAuth();
      });
      
      const passwordInput = screen.getByTestId('password-input');
      await act(async () => {
        fireEvent.change(passwordInput, { target: { value: 'password123' } });
      });
      
      expect(passwordInput).toHaveValue('password123');
    });

    it('應該在輸入時顯示正確的類型', async () => {
      await act(async () => {
        renderAuth();
      });
      
      const emailInput = screen.getByTestId('email-input');
      const passwordInput = screen.getByTestId('password-input');
      
      expect(emailInput).toHaveAttribute('type', 'email');
      expect(passwordInput).toHaveAttribute('type', 'password');
    });
  });

  describe('模式切換', () => {
    it('應該能夠從登入切換到註冊模式', async () => {
      await act(async () => {
        renderAuth();
      });
      
      const toggleButton = screen.getByTestId('toggle-button');
      await act(async () => {
        fireEvent.click(toggleButton);
      });
      
      const submitButton = screen.getByTestId('submit-button');
      expect(submitButton).toHaveTextContent('註冊');
    });
  });
}); 
