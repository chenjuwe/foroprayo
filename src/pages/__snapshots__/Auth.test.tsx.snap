// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Auth Page > should render consistently 1`] = `
<DocumentFragment>
  <div
    class="min-h-screen flex flex-col"
    style="background-color: rgb(244, 228, 221);"
  >
    <div
      class="flex-1 flex flex-col items-center px-6 py-8"
    >
      <div
        class="relative z-10 w-full max-w-sm"
      >
        <div
          class="flex flex-col items-center"
          style="margin-top: 0px;"
        >
          <div
            class="mb-8"
            style="position: absolute; top: 77px; left: 50%; transform: translateX(-50%);"
          >
            <button
              class="text-black"
              style="background-color: rgb(255, 179, 179); width: 100px; height: 30px; border-radius: 15px; font-size: 14px;"
            >
              登入帳號
            </button>
          </div>
          <div
            class="text-black leading-relaxed text-left"
            style="width: 270px; font-size: 14px; position: absolute; top: 158px; left: 50%; transform: translateX(-50%);"
          >
            登入帳號後，代禱紀錄會自動儲存於雲端，
            <br />
            讓每一次代禱都即時同步到你的每一個設備
            <br />
            ，也可以跟好友互相禱告！
          </div>
          <div
            style="margin-top: 254px;"
          >
            <form
              autocomplete="off"
              class="w-full max-w-sm space-y-6 flex flex-col items-center"
            >
              <style>
                
    input::placeholder {
      color: #1694da;
      opacity: 1;
    }
  
              </style>
              <div
                class="relative"
                style="width: 270px;"
              >
                <input
                  autocomplete="off"
                  class="w-full bg-transparent border-b pb-2 text-black focus:outline-none rounded-none"
                  name="email"
                  placeholder="電子信箱"
                  required=""
                  style="font-size: 14px; border-color: rgb(22, 148, 218);"
                  type="email"
                  value=""
                />
              </div>
              <div
                class="relative"
                style="width: 270px;"
              >
                <input
                  autocomplete="new-password"
                  class="w-full bg-transparent border-b pb-2 text-black focus:outline-none rounded-none"
                  name="password"
                  placeholder="輸入密碼"
                  required=""
                  style="font-size: 14px; border-color: rgb(22, 148, 218);"
                  type="password"
                  value=""
                />
              </div>
              <div
                class="flex justify-between items-center pt-4"
                style="width: 270px;"
              >
                <button
                  style="color: rgb(22, 148, 218); font-size: 14px;"
                  type="button"
                >
                  忘記密碼
                </button>
                <button
                  style="color: rgb(22, 148, 218); font-size: 14px;"
                  type="button"
                >
                  註冊帳號
                </button>
              </div>
              <div
                class="space-y-1 text-left"
                style="width: 270px; padding-top: 244px;"
              >
                <div
                  class="text-black"
                  style="font-size: 14px;"
                >
                  App 版本 ＿＿ v1.0.0 ( 2025-11-10 )
                </div>
                <div
                  class="text-black"
                  style="font-size: 14px;"
                >
                  聯絡建議 ＿＿ amen@prayforo.com
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
      <div
        style="position: absolute; top: 544px; left: 50%; transform: translateX(-50%); width: 270px; z-index: 20;"
      >
        <button
          class="text-black w-full"
          style="background-color: rgb(149, 210, 244); height: 30px; border-radius: 15px; font-size: 14px; opacity: 1; margin-bottom: 40px;"
          type="button"
        >
          登入帳號
        </button>
        <a
          class="text-black w-full"
          href="#"
          style="background-color: rgb(77, 155, 217); height: 30px; border-radius: 15px; font-size: 14px; display: flex; align-items: center; justify-content: center; text-decoration: none;"
        >
          訪客帳號
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
